lib/cylc/xtrigger_mgr.py

```diff
@@ -153,89 +162,77 @@ def load_xtrigger_for_restart(self, row_idx, row):
-                try:
-                    val = val % farg_templ
-                except TypeError:
-                    pass
-                args.append(val)
-            for key, val in ctx.func_kwargs.items():
-                try:
-                    val = val % farg_templ
-                except TypeError:
-                    pass
-                kwargs[key] = val
-            ctx.func_args = args
-            ctx.func_kwargs = kwargs
-            ctx.update_command(self.suite_source_dir)
+            ctx = self.get_xtrig_ctx(itask, label)
             sig = ctx.get_signature()
             if sigs_only:
                 res.append(sig)
             else:
                 res.append((label, sig, ctx, satisfied))
         return res
 
+    def get_xtrig_ctx(self, itask, label):
+        """Get a real function context from the template."""
+        farg_templ = {
+            TMPL_TASK_CYCLE_POINT: str(itask.point),
+            TMPL_TASK_NAME: str(itask.tdef.name),
+            TMPL_TASK_IDENT: str(itask.identity)
+        }
+        farg_templ.update(self.farg_templ)
+        ctx = deepcopy(self.functx_map[label])
+        ctx.point = itask.point
+        kwargs = {}
+        args = []
+        for val in ctx.func_args:
+            try:
+                val = val % farg_templ
+            except TypeError:
+                pass
+            args.append(val)
+        for key, val in ctx.func_kwargs.items():
+            try:
+                val = val % farg_templ
+            except TypeError:
+                pass
+            kwargs[key] = val
+        ctx.func_args = args
+        ctx.func_kwargs = kwargs
+        ctx.update_command(self.suite_source_dir)
+        return ctx
+
     def collate(self, itasks):
         """Get list of all current xtrigger sigs."""
         self.all_xtrig = []
-        self.all_xclock = []
         for itask in itasks:
-            self.all_xtrig += self._get_xtrig(itask, sigs_only=True)
-            if itask.state.xclock is not None:
-                self.all_xclock.append(self._get_xclock(itask, sig_only=True))
+            self.all_xtrig += self._get_xtrigs(itask, sigs_only=True)
 
     def satisfy_xtriggers(self, itask, proc_pool):
         """Attempt to satisfy itask's xtriggers."""
-        for label, sig, ctx, _ in self._get_xtrig(itask, unsat_only=True):
+        for label, sig, ctx, _ in self._get_xtrigs(itask, unsat_only=True):
+            if sig.startswith("wall_clock"):
+                # Special case: synchronous clock check.
+                ctx.func_kwargs.update(
+                    {
+                        'point_as_seconds': itask.get_point_as_seconds(),
+                    }
+                )
+                if wall_clock(*ctx.func_args, **ctx.func_kwargs):
+                    itask.state.xtriggers[label] = True
+                    self.sat_xtrig[sig] = {}
+                    LOG.info('clock xtrigger satisfied: %s = %s' % (
+                        label, str(ctx)))
```

Comment:
Simpler:

```python
LOG.info('clock xtrigger satisfied: %s = %s', label, ctx)
```

Reply:
:+1:

Reply:
(In other words, I wonder why logging has its own string templating functionality?)

Reply:
Lazy evaluation. The interface is roughly `.info(formatstr, *formatargs)`, right? The logger does the `formatstr % formatargs` operation only when it *needs* to emit the message.

Reply:
A-ha, right, excellent point.  In that case I will use it :+1: